### ISSUE TITLE
Hang formatting nested functions

### ISSUE THREAD

**Issue by gabrielmaldi** [prettier/prettier#1250]
Hey, I just found that prettier hangs when formatting this:

```js
someObject.someFunction().then(function () {
    return someObject.someFunction().then(function () {
        return someObject.someFunction().then(function () {
            return someObject.someFunction().then(function () {
                return someObject.someFunction().then(function () {
                    return someObject.someFunction().then(function () {
                        return someObject.someFunction().then(function () {
                            return someObject.someFunction().then(function () {
                                return someObject.someFunction().then(function () {
                                    anotherFunction();
                                });
                            });
                        });
                    });
                });
            });
        });
    });
});
```

(I'm embarrassed to have this piece of code in a codebase, it's because of a limitation of a third-party library 😄)

If I remove two of the innermost nested calls, prettier does end and format correctly (it takes a while, though). So it seems that it's not a hang but rather just taking exponentially forever when dealing with this (messed up) code.

I'm running prettier `1.0.2` on macOS 10.12.4 through the CLI: `prettier file.js`

Thanks

**Comment by vjeux** [comment]
Oh wow, great catch. We ran prettier through a ton of code and couldn't make it break like this. cc @jlongster

**Comment by vjeux** [comment]
Looks like this is generating an enormous amount of IR. I kept only 4 levels of indentation and when trying to print it go

```js
<--- Last few GCs --->

[44247:0x102801600]    70984 ms: Mark-sweep 1400.4 (1509.1) -> 1400.4 (1478.1) MB, 1035.0 / 0.0 ms  (+ 0.0 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 1035 ms) last resort
[44247:0x102801600]    72005 ms: Mark-sweep 1400.4 (1478.1) -> 1400.3 (1478.1) MB, 1019.7 / 0.0 ms  last resort


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x498588266a1 <JS Object>
    1: printArrayItems(aka printArrayItems) [/Users/vjeux/random/prettier/src/printer.js:~3513] [pc=0x217385a5b36](this=0x2eb042382311 <undefined>,path=0x37406bb44059 <a FastPath with map 0x2a9f209139d1>,options=0x12ca1a13cd41 <an Object with map 0x2a9f2090b681>,printPath=0xc6e5b075de9 <String[8]: elements>,print=0x37406bb440b9 <JS Function printGenerically (SharedFunctionInfo 0x26b50a4faed1)>)
...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/usr/local/bin/node]
 5: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
 6: 0x2173820ed46
Abort trap: 6
```

**Comment by gabrielmaldi** [comment]
FWIW, strangely if I remove all `someObject.` occurrences and leave it like this:

```js
someFunction().then(function () {
    return someFunction().then(function () {
        return someFunction().then(function () {
            return someFunction().then(function () {
                return someFunction().then(function () {
                    return someFunction().then(function () {
                        return someFunction().then(function () {
                            return someFunction().then(function () {
                                return someFunction().then(function () {
                                    anotherFunction();
                                });
                            });
                        });
                    });
                });
            });
        });
    });
});
```

prettier takes less than 1 second to finish.

**Comment by vjeux** [comment]
One thing we should probably do is add a safeguard and if we are going to be in those degenerative cases, then just take the expanded version.

**Comment by vjeux** [comment]
```js
someFunction().then(function () {
    return someFunction().then(function () {
        return someFunction().then(function () {
            return someFunction().then(function () {
                return someFunction().then(function () {
                    return someFunction().then(function () {
                        return someFunction().then(function () {
                            return someFunction().then(function () {
                                return someFunction().then(function () {
                                    anotherFunction();
                                });
                            });
                        });
                    });
                });
            });
        });
    });
});
```

has 9841 conditionalGroups.

**Comment by vjeux** [comment]
In order to get you unblocked, you can put a `// prettier-ignore` on the line before.

```js
// prettier-ignore
someObject.someFunction().then(function () {
    return someObject.someFunction().then(function () {
        return someObject.someFunction().then(function () {
            return someObject.someFunction().then(function () {
                return someObject.someFunction().then(function () {
                    return someObject.someFunction().then(function () {
                        return someObject.someFunction().then(function () {
                            return someObject.someFunction().then(function () {
                                return someObject.someFunction().then(function () {
                                    anotherFunction();
                                });
                            });
                        });
                    });
                });
            });
        });
    });
});
```

which is going to print it as is and prettier is going to be instant.

**Comment by tunnckoCore** [comment]
it's not a bug for me. that code is not serious and is totally ridiculous. you shoul separate it into separate functions and can be banned with 2 eslint rules (im calling it no nested hell)

**Comment by vjeux** [comment]
@tunnckoCore prettier doesn't make assumptions in how you actually write your program, it is going to format it. The problem here is that there is some exponential aspect inside of prettier that triggers here, so we should ensure that it doesn't happen.

**Comment by jlongster** [comment]
I'm honestly surprised we haven't hit an edge case like this earlier, since we don't some pretty crazy AST traversal in special cases (like member expressions).

Thanks for the bug! I don't have time to look right now but we'll profile it and see what we can do. It makes me a little happy that, after running prettier on probably millions of lines of code by this point, we haven't seen code like this before. :)

**Comment by vjeux** [comment]
@jlongster the problem here is the classic conditionalGroup explosion. Both MemberExpression and CallExpression are doing one. So it's clearly a case that has O(n^2) runtime behavior.

**Comment by jeresig** [comment]
I actually just hit this same issue today! My code is similarly embarrassing, haha:
https://gist.github.com/jeresig/11002bba056f2dc9ccc64577b4814ed9

If I put that code in a separate file it seems to just hang. If I add `// prettier-ignore` at the beginning it finishes, although it still takes about 1700ms. I can use the ignore for now, at the very least!

This is with prettier v1.1.0 on macOS 10.12.3 through the cli.

**Comment by vjeux** [comment]
I bisected the issue and there are two culprits:

#927 makes it go from instant to taking a few seconds.
#1217 makes it go from taking a few seconds to not terminating.

I'm trying to figure out what we can do to mitigate this. So sorry about this!

**Comment by jeresig** [comment]
@vjeux No worries at all! Thank you so much for investigating! (prettier is amazing, I'm absolutely loving it!)

**Comment by vjeux** [comment]
I'm currently on weekend outside my house so won't fix it until tuesday but I have two ideas to solve this.

For 1), we should only call print once with expandArg and have the call return two options instead of printing it twice incurring a O(n^2) behavior. This is likely going to be a bit tricky with genericPrintNoParens that does some magic.

For 2), we should be able to use a weakmap to memoize the computation instead of doing it all the time. I'm not 100% convinced yet that the issue is around caching and not actually doing too much work but worth a shot.

**Comment by gabrielmaldi** [comment]
@vjeux @jlongster thank you for being so receptive, responsive, and making prettier great!

(One day I'd ❤️ to have prettier format my HTML — following its two core principles: parse and re-print, acknowledge maximum line length — as well 😃).

**Comment by SimenB** [comment]
Html and css can be (partially) accomplished today using https://github.com/beautify-web/js-beautify/blob/master/README.md#css--html

But one meta thing calling all these tools (including stylefmt and provably others I don't know exists) would be truly amazing :smile:

**Comment by gabrielmaldi** [comment]
@SimenB I've been using JS Beautifier for a while now 😄 But IMO prettier is the best formatter because it parses and prints ignoring all the original formatting, making the output 100% consistent. I'd love to have this not only for JavaScript but also HTML. (Sorry for being off-topic).
